core/src/main/java/org/gbif/pipelines/core/ws/client/internal/GBIFInternalServiceClient.java

```diff
@@ -0,0 +1,174 @@
+  }
+
+  /**
+   * requests http://api.gbif.org/v1/dataset/{datasetid}
+   *
+   * @return dataset info for provided dataset id
+   */
+  public JsonObject getDatasetInfo(String datasetUUID) {
+    Objects.requireNonNull(datasetUUID);
+    return performCall(service.getDatasetInfo(datasetUUID)).getAsJsonObject();
+  }
+
+  /**
+   * requests http://api.gbif.org/v1/installation/{installation_id}
+   *
+   * @return installation info
+   */
+  public JsonObject getInstallationInfo(String installationUUID) {
+    Objects.requireNonNull(installationUUID);
+    return performCall(service.getInstallationInfo(installationUUID)).getAsJsonObject();
+  }
+
+  /**
+   * perform webservice call to aggregate the needed GBIF terms
+   *
+   * @param datasetUUID datasetUUID
+   *
+   * @return aggregated GBIFTerms response for the provided datasetUUID
+   */
+  public GBIFInternalResponse getInternalResponse(String datasetUUID) throws ExecutionException {
+    Objects.requireNonNull(datasetUUID, "DatasetUUID cannot be null");
+    //create callable on cache miss the GBIFInternalResponse for the dataset id is stored in the cache
+    Callable<GBIFInternalResponse> callableResponse = () -> {
+      GBIFInternalResponse response = new GBIFInternalResponse();
+      response.setDatasetKey(datasetUUID);
+
+      JsonObject dataset = getDatasetInfo(datasetUUID);
+
+      Optional.ofNullable(dataset.getAsJsonPrimitive(DATASET_TITLE_KEY))
+        .ifPresent((title) -> response.setDatasetTitle(title.getAsString()));
+
+      Optional.ofNullable(dataset.getAsJsonPrimitive(PUB_ORGANIZATION_KEY)).ifPresent((orgKey) -> {
+        JsonObject orgInfo = getOrganizationInfo(orgKey.getAsString());
+        response.setPublishingOrgKey(orgKey.getAsString());
+        Optional.ofNullable(orgInfo.getAsJsonPrimitive(PUB_ORG_COUNTRY_KEY))
+          .ifPresent((countryObject) -> response.setPublishingCountry(countryObject.getAsString()));
+      });
+
+      Optional.ofNullable(dataset.getAsJsonPrimitive(INSTALLATION_KEY)).ifPresent((key) -> {
+        JsonObject installationInfo = getInstallationInfo(key.getAsString());
+        Optional.ofNullable(installationInfo.getAsJsonPrimitive(INSTALLATION_TYPE_KEY))
+          .ifPresent((type) -> response.setProtocol(type.getAsString()));
+      });
+
+      JsonArray networks = getNetworkFromDataset(datasetUUID);
+      List<String> networkKeys = new ArrayList<>(networks.size());
+      networks.iterator()
+        .forEachRemaining((element) -> networkKeys.add(element.getAsJsonObject()
+                                                         .get(DATASET_NETWORK_KEY)
+                                                         .getAsString()));
+      response.setNetworkKey(networkKeys);
+      return response;
+    };
+    return datasetResponseCache.get(datasetUUID, callableResponse);
+  }
+
+  /**
+   * executes request and handles response and errors.
+   */
+  private JsonElement performCall(Call<JsonElement> serviceCall) {
```

Comment:
IOException should be propagated here, remove all exception handling and leave that responsibility to the caller

Reply:
In that case i will be ending up rewriting this code in 4 different places when i am calling this function , You want me to do it ?